website/blog: publish I gambled against React and lost (and I don’t regret a thing) (#5482)
goauthentik/authentik: website/blog/2023-05-04-i-gambled-against-react-and-lost/item.md

So it was with React.

In this post, I’ll explain why I made this decision, how it did and didn’t pay off, and why, ultimately, I don’t regret it. The point isn’t to sway you toward or away from React or to make an argument about web frameworks in general, but to encourage a discussion about the choices early-stage startups have to make.

<!--truncate-->

And when I say startups here, I’m not talking about the scale-ups still calling themselves startups or the startup darlings with more dollars of venture capital than they can burn through, I’m talking about people like me: developers working full-time on trying to build a company out of a popular project.

## React vs. Lit: A brief history and a one-sided victory

React was first released in 2013 and when I was building the initial open source project that would eventually become Authentik Security, React was a known option but not, as it arguably is now, a default.

In 2015, Netflix and Airbnb started using React. Reactive Native for iOS came out that same year and later, React Native for Android. By 2018, the year I made the decision, 28.3% of developers had started using React.

> 

Reviewer: I'd suggest you add a caption here, to clearly label the chart as showing 2018 data. We wouldn't want people to read to quickly and miss that fact and get all twisted. Ditto on chart from 2022.

![image2.png](image2.png)

React was popular and rising but not an obvious choice – not a choice that anyone would be surprised at for not choosing. Now, of course, it’s all different.

2022 StackOverflow [research](https://survey.stackoverflow.co/2022/) shows that 44.31% of developers now use React.

![image3.png](image3.png)

> To get a sense of Lit’s popularity, in comparison, I had to narrow my research to JavaScript developers. Among JS developers, according to the [2022 State of JavaScript survey](https://stackdiary.com/front-end-frameworks/), Lit is used by only 6%.

Reviewer: ```suggestion
To get a sense of Lit’s popularity, in comparison, I had to narrow my search to JavaScript developers. Among JS developers, according to the [2022 State of JavaScript survey](https://stackdiary.com/front-end-frameworks/), Lit is used by only 6%.
```

![image1.png](image1.png)

It’s not a sheer popularity difference:

-   React has 232,063 packages on [NPM](https://www.npmjs.com/search?q=react) (Lit has [2451](https://www.npmjs.com/search?q=lit))
-   React has over [1600 contributors](https://github.com/facebook/react) (Lit has [149](https://github.com/lit/lit)).
-   React has over [14,000,000 users](https://github.com/facebook/react). (Lit has over [26,000](https://github.com/lit/lit)).

> I won’t say this success was impossible to predict but it also wasn’t easy.

Reviewer: ```suggestion
I won’t say this success was impossible to predict but it also wasn’t obvious.
```

### Why I chose Lit

The most important context to this decision is in the header: Authentik Security is now a team and a company, but in 2018, Authentik Security was just me and Authentik Security was authentik, the open source project.

I was building authentik as a hobby and all the work I did on it was done outside my full-time work. In that period, a few limitations rose to the top:

-   I had few resources, both in terms of time and funds.
-   I wasn’t a front-end engineer.
-   I wanted to be able to build features over time, at my own speed, rather than work on nothing else for three months before having anything to show.

At the time, hiring potential wasn’t on my mind: Building and shipping was.

In the beginning, the project was entirely rendered server-side. With Django, I had some tooling available that made it easy to build forms for submitting and showing data as well as adding pagination, tables, and sorting.

But a few features emerged on the roadmap that soon made this setup feel less simple. I wanted to give users the ability to refresh data with a button press and complete searches without reloading the page, for example.

> In late 2020, I started actually migrating to Lit, already knowing a lot of work was ahead of myself. At that point it would often happen that when new features were introduced, they would be in the UI but not in the API, often due to forgetfulness. This turned things into a constant cat and mouse game to keep the API updated without knowing what endpoints were actually required.

Reviewer: ```suggestion
In late 2020, I started migrating to Lit, already knowing a lot of work was ahead of myself. At that point it would often happen that when new features were introduced, they would be in the UI but not in the API, often due to forgetfulness. This turned things into a constant cat and mouse game to keep the API updated without knowing what endpoints were actually required.
```

> Then, toward the end of 2022, with support from Open Core Ventures, [authentik became Authentik Security](https://goauthentik.io/blog/2022-11-02-the-next-step-for-authentik). And once again, the product roadmap made some demands.

Reviewer: ```suggestion
Then, toward the end of 2022, with support from [Open Core Ventures](https://opencoreventures.com/), [authentik became Authentik Security](https://goauthentik.io/blog/2022-11-02-the-next-step-for-authentik). And once again, the product roadmap made some demands.
```

> Inflexibility was an early Authentik flaw. Because we relied on server-side rendering, we also relied – and made our users rely – on browsers. But we had always wanted to go beyond that, to get to a point where there was a dynamic back-and-forth between the client and the server.

Reviewer: ```suggestion
Inflexibility was an early flaw in authentik. Because we relied on server-side rendering, we also relied – and made our users rely – on browsers. But we had always wanted to go beyond that, to get to a point where there was a dynamic back-and-forth between the client and the server.
```

If we wanted, for example, to add a new component for a table, we could have the server return it without changing too much on the server side. But on the client side, we could add logic and pagination and other features over time. That just wasn’t doable before we adopted Lit and Lit made that kind of migration path more doable than React would have.

## Tradeoffs all the way down

Scroll through HackerNews and you’ll find numerous, often tense or fiery, discussions of frameworks. Some people will be diligent about costs and benefits but many will be brand-loyal; arguing on the behalf of one framework and trashing all the others.

> “My fellow engineers, please stop asking ‘Is this process good or bad?’ and start asking “Is it well-suited to my situation?” - Jocelyn Goldfein, former engineering executive at Facebook and VMware

The lesson I learned from choosing between Lit and React – and living with the consequences – is that these decisions are tradeoffs all the way down.

> Jocelyn Goldfein, a former engineering executive at Facebook and VMware, [frames it well](https://review.firstround.com/the-right-way-to-ship-software): “My fellow engineers, please stop asking ‘Is this process good or bad?’ and start asking “Is it well-suited to my situation?”

Reviewer: Remove this paragraph and leave quote above...

Author: ```suggestion
```

The tricky part of this question is that identifying your situation is harder than it seems. Scale-ups and enterprises can rely on better pattern-matching than true, early startups can. Each startup, like [unhappy families](https://www.goodreads.com/quotes/7142-all-happy-families-are-alike-each-unhappy-family-is-unhappy), is unhappy in its own way.

When I defined our situation, three factors came to the top:

1. Shipping was our top priority.
2. A faster shift in rendering approaches was more important than slower hiring speeds.
3. Flexibility was more valuable than the help a framework would offer.

Remember, at the very beginning, it was me alone working on an open source hobby project when I had time outside of a full-time job. Especially considering the work I had already done and my lack of frontend engineering skills, it was clear that it was more important to build and ship than to restart.

> Here, the tradeoff was that I was able to build Authentik faster and deliver it to users sooner, but now I’m missing out on the advantages that being involved with a thriving community like React would offer.

Reviewer: ```suggestion
Here, the tradeoff was that I was able to build authentik faster and deliver it to users sooner, but now I’m missing out on the advantages that being involved with a thriving community like React would offer.
```

With Lit, we could migrate more easily and more practically, migrating more and more over time rather than either rebuilding or migrating all at once. Looking back, I don’t know if I could have completed such a major shift in how we were rendering and if I could have, I don’t know if it would even be done by now.

> Here, the tradeoff is that finding and hiring developers with Lit experience is much harder than finding and hiring developers with React experience. Hiring developers is already notoriously hard and we made it harder. There is however the advantage that due to the smaller userbase of Lit, that in the rare case we do find a Lit developer, they tend to be very passionate about the same reasons why we chose Lit.

Reviewer: ```suggestion
So Yes, finding and hiring developers with Lit experience is much harder than finding and hiring developers with React experience. Hiring developers is already notoriously hard and we made it harder. There is however the advantage that due to the smaller user-base of Lit, that in the rare case we do find a Lit developer, they tend to be very passionate about the same reasons why we chose Lit.
```
